river/ensemble/streaming_random_patches.py

```diff
@@ -0,0 +1,474 @@
+            if k != 0 and k < n_features:
+                # For low dimensionality it is better to avoid more than
+                # 1 classifier with the same subspace, thus we generate all
+                # possible combinations of subsets of features and select
+                # without replacement.
+                # n_features is the total number of features and k is the
+                # actual size of a subspace.
+                if n_features <= 20 or k < 2:
+                    if k == 1 and n_features > 2:
+                        k = 2
+                    # Generate n_models subspaces from all possible
+                    # feature combinations of size k
+                    self._subspaces = []
+                    for i, combination in enumerate(itertools.cycle(
+                            itertools.combinations(features, k))):
+                        if i == self.n_models:
+                            break
+                        self._subspaces.append(list(combination))
+                # For high dimensionality we can't generate all combinations
+                # as it is too expensive (memory). On top of that, the chance
+                # of repeating a subspace is lower, so we randomly generate
+                # subspaces without worrying about repetitions.
+                else:
+                    self._subspaces = [random_subspace(all_features=features, k=k, rng=self._rng)
+                                       for _ in range(self.n_models)]
+            else:
+                # k == 0 or k > n_features (subspace size is larger than the
+                # number of features), then default to re-sampling
+                self.training_method = self._TRAIN_RESAMPLING
+
+    def _init_ensemble(self, features: list):
+        self._generate_subspaces(features=features)
+
+        subspace_indexes = np.arange(self.n_models)  # For matching subspaces with ensemble members
+        if self.training_method == self._TRAIN_RANDOM_PATCHES or \
+                self.training_method == self._TRAIN_RANDOM_SUBSPACES:
+            # Shuffle indexes
+            self._rng.shuffle(subspace_indexes)
+
+        # Initialize the ensemble
+        for i in range(self.n_models):
+            # If self.training_method == self._TRAIN_RESAMPLING then subspace is None
+            subspace = self._subspaces[subspace_indexes[i]]
+            self.models.append(self._base_learner_class(
+                idx_original=i,
+                base_model=self.model,    # TODO replace with clone
+                metric=copy.deepcopy(self.metric),
+                created_on=self._n_samples_seen,
+                base_drift_detector=self.drift_detector,
+                base_warning_detector=self.warning_detector,
+                is_background_learner=False,
+                features=subspace,
+                nominal_attributes=self.nominal_attributes,
+                rng=self._rng))
+
+    def reset(self):
```

Comment:
Is there need for a bespoke `reset` method, or can we handle this generically with a function in `utils`?

Reply:
I left it for reference. If we go with a generic function then we can remove it.